letsencrypt/cli.py

```diff
@@ -1051,6 +1030,51 @@ def _plugins_parsing(helpful, plugins):
 
     helpful.add_plugin_args(plugins)
 
+    # These would normally be a flag within the webroot plugin, but because
+    # they are parsed in conjunction with --domains, they live here for
+    # legibiility. helpful.add_plugin_ags must be called first to add the
+    # "webroot" topic
+    helpful.add("webroot", "-w", "--webroot-path", action=WebrootPathProcessor,
+                help="public_html / webroot path")
+    parse_dict = lambda s: dict(json.loads(s))
+    helpful.add("webroot", "--webroot-map", default={}, type=parse_dict,
+                help="Mapping from domains to webroot paths")
+
+
+class WebrootPathProcessor(argparse.Action): # pylint: disable=missing-docstring
+    def __call__(self, parser, config, webroot, option_string=None):
+        """
+        Keep a record of --webroot-path / -w flags during processing, so that
+        we know which apply to which -d flags
+        """
+        if not config.webroot_path:
+            config.webroot_path = []
+            # if any --domain flags preceded the first --webroot-path flag,
+            # apply that webroot path to those; subsequent entries in
+            # config.webroot_map are filled in by cli.DomainFlagProcessor
+            if config.domains:
+                config.webroot_map = dict([(d, webroot) for d in config.domains])
+            else:
+                config.webroot_map = {}
+        config.webroot_path.append(webroot)
+
+
+class DomainFlagProcessor(argparse.Action): # pylint: disable=missing-docstring
```

Comment:
As I said briefly in IRC, I don't think that the argparsing for domains should do anything with `webroot-path`. This also goes against our architecture of having the plugins do their argument parsing.

I think that changes to domain parsing can be reverted and logic similar to below can be used in `webroot.py` to achieve the same effect. There are a couple ways to do this. Basically, you need to keep track of the previous `webroot-path` flag. Then when you encounter a new `webroot-path` flag, you know you will use that webroot for all new domains found in `config.domains`. The very last `webroot-path` flag will require special treatment and will have to be applied outside of argument parsing, but I believe this can be done nicely using a function to avoid duplicated code.

This relies on `argparse` parsing arguments in the order they appear on the command line, but so does the current code in this PR.